library/core/qsl-base/src/main/java/org/quiltmc/qsl/api/event/Event.java

```diff
@@ -0,0 +1,118 @@
+ * pass an instance of {@code T} into {@link #register}.
+ *
+ * <pre><code>
+ * public static final Event&lt;Runnable&gt; EXAMPLE = ...; // implementation
+ *
+ * public void registerEvents() {
+ *     // Since runnable is a functional interface, we can use the lambda form.
+ *     EXAMPLE.register(() -> {
+ *         // Do something
+ *     });
+ *
+ *     // Or we can use a method reference.
+ *     EXAMPLE.register(this::runSomething);
+ *
+ *     // Or implement the callback onto a class.
+ *     // You can also use an anonymous class here; for brevity that is not included.
+ *     EXAMPLE.register(new ImplementedOntoClass());
+ * }
+ *
+ * public void runSomething() {
+ *     // Do something else
+ * }
+ *
+ * // When implementing onto a class, the class must implement the same type as the event invoker.
+ * class ImplementedOntoClass implements Runnable {
+ *     public void run() {
+ *         // Do something else again
+ *     }
+ * }
+ * </code></pre>
+ *
+ * <h2>Example: Executing an event</h2>
+ *
+ * Executing an event is done by calling a method on the event invoker. Where {@code T} is Runnable, executing an event
+ * is done through the following:
+ *
+ * <pre><code>
+ * EXAMPLE.invoker().run();
+ * </code></pre>
+ *
+ * <h2>Best practices for implementations of {@link Event}</h2>
+ *
+ * Implementations do not need to follow all the best practices, but you should consider providing the following:
+ * <ul>
+ *     <li>Synchronization when registering callbacks.
+ *     <ul>
+ *         <li>Registration of callbacks may occur on multiple threads, so thread safety when registering callbacks may
+ *         be useful.
+ *         <li>When the invoker is being executed, a thread safe implementation should ensure the current invoker's
+ *         state should not change. This means a newly registered invoker should only be visible for execution after
+ *         any current open executions of the invoker finish.
+ *     </ul>
+ * </ul>
+ *
+ * @param <T> the type of the invoker used to execute an event
+ */
+public abstract class Event<T> {
+	/**
+	 * The invoker field.
+	 *
+	 * <p>This field is volatile to allow implementations of this class to use a suitable thread safety mechanism if
+	 * desired.
+	 */
+	private volatile T invoker;
+
+	/**
+	 * Returns the invoker instance.
+	 *
+	 * <p>This method is used to invoke an event and execute hidden callbacks implemented by the invoker.
+	 *
+	 * @return the invoker instance
+	 */
+	public final T invoker() {
+		return this.invoker;
+	}
+
+	/**
+	 * Sets the invoker instance.
+	 *
+	 * <p>This method is protected and final since it is only intended for implementors of the {@link Event} class to
+	 * set the invoker instance.
+	 *
+	 * <p>Since the invoker field is volatile, the invoker field should be available immediately for use.
+	 *
+	 * @param invoker the invoker
```

Comment:
```suggestion
	 * @param invoker the invoker instance
```

For consistency with the getter.